### PR TITLE
feat: support custom token

### PR DESCRIPTION
**Pull request by MadCcc**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link
close https://github.com/ant-design/ant-design/issues/36073

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Support third party customized token          |
| 🇨🇳 Chinese |  支持三方库自定义 token         |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/_util/theme/index.tsx [hunk 1]
```diff
@@ -93,3 +93,35 @@ export type GenerateStyle<
   ComponentToken extends object = AliasToken,
   ReturnType = CSSInterpolation,
 > = (token: ComponentToken) => ReturnType;
+
+export type CustomTokenOptions<
+  CustomSeedToken extends Record<string, any>,
+  CustomOverride extends object,
+> = {
+  seedToken?: CustomSeedToken;
+  override?: CustomOverride;
```
Comment (zombieJ) [inline components/_util/theme/index.tsx:102]: Note: 先不用 ComponentToken

File: components/config-provider/__tests__/token.test.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  };
+
+  const useMyToken: UseToken<AliasToken & MySeedToken & MyAliasToken> = () =>
+    useCustomToken<MySeedToken, MyAliasToken>({
+      seedToken: { colorAlert: 'red' },
+      formatToken: derivativeToken => ({
+        ...derivativeToken,
+        colorAlertActive: 'purple',
+      }),
+      hashed: true,
```
Comment (zombieJ) [inline components/config-provider/__tests__/token.test.tsx:26]: 这个自动掉就好~

File: components/config-provider/__tests__/token.test.tsx [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    };
+
+    const { container } = render(<Demo />);
+    const finalToken = JSON.parse(container.firstElementChild?.innerHTML ?? '');
+    expect(finalToken).toHaveProperty('colorAlert', 'red');
+    expect(finalToken).toHaveProperty('colorAlertActive', 'purple');
+  });
+
+  it('registerToken', () => {
+    const useStyle = genStyleHook(token => ({
```
Comment (zombieJ) [inline components/config-provider/__tests__/token.test.tsx:44]:
```tsx
// ✅
function useMyToken() {
  const myToken = useCustomToken({ ... });
  return myToken;
}
```

```tsx
// Customize register
function useMyStyleRegister(componentName, styleFn) {
  const myToken = useMyToken();
  return useStyleRegister(myToken, styleFn);
}
```

```tsx
const Button = () => {
  useMyStyleRegister('myButton', (token) => CSSObject);
};
```

File: components/config-provider/index.tsx [hunk 4]
```diff
@@ -302,5 +306,8 @@ ConfigProvider.ConfigContext = ConfigContext;
 ConfigProvider.SizeContext = SizeContext;
 ConfigProvider.config = setGlobalConfig;
 ConfigProvider.useToken = useToken;
+ConfigProvider.useStyle = useStyle;
+ConfigProvider.useCustomToken = useCustomToken;
+ConfigProvider.useCustomStyle = useCustomStyle;
```
Comment (afc163) [inline components/config-provider/index.tsx:311]: 感觉有点复杂，感觉 antd 只暴露出一个 useStyle 就够了。现在 API 太多用户不知道怎么选

**Comment by jfwong** [comment]
有谁可以告诉我， 在以上的preview里面，那个different的功能是怎样实现的吗？
我想在我的项目里边实现这个功能，这样在验证UI改动的时候可以很简单的分辨出来

**Pull request merged**